kmk/extensions/ble_split.py

```diff
@@ -0,0 +1,195 @@
+        self._is_target = True
+        self._uart_buffer = []
+        self.hid_type = hid_type
+        self.split_flip = split_flip
+        self.split_side = split_side
+        self.split_offset = None
+        self._ble = BLERadio()
+        self._ble_last_scan = ticks_ms() - 5000
+        self._is_target = True
+        self._connection_count = 0
+        self._uart = None
+        self._uart_connection = None
+        self._advertisment = None
+        self._advertising = False
+        self._uart_interval = uart_interval
+        self._psave_enable = False
+
+    def __repr__(self):
+        return f'BLE_SPLIT({self._to_dict()})'
+
+    def _to_dict(self):
+        return f'''
+        BLE_Split( _ble={self._ble}
+        _ble_last_scan={self._ble_last_scan}
+        _is_target={self._is_target}
+        _uart_buffer={self._uart_buffer}
+        _split_flip={self.split_flip}
+        _split_side={self.split_side} )
+        '''
+
+    def on_runtime_enable(self, keyboard):
+        return
+
+    def on_runtime_disable(self, keyboard):
+        return
+
+    def during_bootup(self, keyboard):
+        self._ble.name = str(getmount('/').label)
+        if self._ble.name.endswith('L'):
+            # If name ends in 'L' assume left and strip from name
+            self._is_target = True
+        elif self._ble.name.endswith('R'):
+            # If name ends in 'R' assume right and strip from name
+            self._is_target = False
+        else:
+            self._is_target = bool(self.split_side == 'Left')
+
+        if self.split_flip and not self._is_target:
+            keyboard.col_pins = list(reversed(keyboard.col_pins))
+
+        self.split_offset = len(keyboard.col_pins)
+
+        # Attempt to sanely guess a coord_mapping if one is not provided.
+        if not keyboard.coord_mapping:
+            keyboard.coord_mapping = []
+
+            rows_to_calc = len(keyboard.row_pins) * 2
+            cols_to_calc = len(keyboard.col_pins) * 2
+
+            for ridx in range(rows_to_calc):
+                for cidx in range(cols_to_calc):
+                    keyboard.coord_mapping.append(intify_coordinate(ridx, cidx))
+
+    def before_matrix_scan(self, keyboard):
+        self._check_all_connections()
+        return self._receive()
+
+    def after_matrix_scan(self, keyboard, matrix_update):
+        if matrix_update:
+            matrix_update = self._send(matrix_update)
+            return matrix_update
+        return None
+
+    def before_hid_send(self, keyboard):
+        return
+
+    def after_hid_send(self, keyboard):
+        return
+
+    def on_powersave_enable(self, keyboard):
+        if self._uart_connection and not self._psave_enable:
+            self._uart_connection.connection_interval = self._uart_interval
+            self._psave_enable = True
+
+    def on_powersave_disable(self, keyboard):
+        if self._uart_connection and self._psave_enable:
+            self._uart_connection.connection_interval = 11.25
```

Comment:
We can move it, but that is a standard that would not likely change. That is the minimum speed that BLE HID devices can act on spec. The variable number is set for powersave changes, but this number being defined in a spec is why it's hard coded and not moved to a variable using more memory for no reason.

Reply:
FWIW that's not how numbers in Python work - the RAM would be similar either way, as the number has to be allocated either way, and since it's being assigned to a class instance that outlives the function being executed, it will guaranteed go onto the heap instead of the stack. I won't speak to whether the current implementation uses more RAM and/or allocation cycles or not, because that gets into quirks of the CPython VM that I don't know whether they translate to MicroPython and don't care enough right now to look up (however, if you're ever up for reading C, I'd recommend reading the relevant MicroPython source - it's a pretty approachable codebase that's reasonable to understand!)

Reply:
I'm willing to put that into a variable, though that number should not be touched until the BLE spec changes. My reason was mostly so it would not be touched.